crates/analyzer/src/traversal/expressions.rs

```diff
@@ -873,15 +880,73 @@ fn check_for_call_to_init_fn(
     }
 }
 
+fn check_for_self_param(scope: &mut BlockScope, use_span: Span) -> Result<(), FatalError> {
+    if scope.root.function.signature(scope.db()).self_decl == SelfDecl::None {
+        scope.fancy_error(
+            "`self` is not defined",
+            vec![Label::primary(use_span, "undefined value")],
+            vec![
+                "add `self` to the scope by including it in the function signature".to_string(),
+                "Example: `fn foo(self, bar: bool)`".to_string(),
+            ],
+        );
+        Err(FatalError::new())
+    } else {
+        Ok(())
+    }
+}
+
 fn expr_call_self_attribute(
+    scope: &mut BlockScope,
+    func_name: &str,
+    name_span: Span,
+    self_span: Span,
+    args: &Node<Vec<Node<fe::CallArg>>>,
+) -> Result<ExpressionAttributes, FatalError> {
+    check_for_call_to_init_fn(scope, func_name, name_span)?;
+    check_for_self_param(scope, self_span)?;
+
+    if let Some(func) = scope.root.self_contract_function(func_name) {
+        let sig = func.signature(scope.root.db);
+        validate_arg_count(scope, func_name, name_span, args, sig.params.len());
+        validate_arg_types(scope, func_name, args, &sig.params)?;
```

Comment:
This is recreating a bug that was fixed recently in #518

Here's a snippet that doesn't cause a compile error with this change but should:

```
contract Foo:

    pub fn baz(self):
        self.bar(doesnt_exist=1, me_neither=2)
    
    pub fn bar(self, val1: u256, val2: u256):
        pass
```

We should add this as a test case variant of the `mislabeled_call_args.fe` fixture that we already have (but which doesn't apply anymore since it doesn't use `self`)